crates/consensus-logic/src/csm/worker.rs

```diff
@@ -0,0 +1,372 @@
+///
+/// Unable to be shared across threads.  Any data we want to export we'll do
+/// through another handle.
+#[allow(unused)]
+pub struct WorkerState<D: Database> {
+    /// Consensus parameters.
+    params: Arc<Params>,
+
+    /// CSM worker config, *not* params.
+    config: CsmExecConfig,
+
+    /// Underlying database hierarchy that writes ultimately end up on.
+    // TODO should we move this out?
+    database: Arc<D>,
+
+    /// L2 block manager.
+    l2_block_manager: Arc<L2BlockManager>,
+
+    /// Checkpoint manager.
+    checkpoint_manager: Arc<CheckpointDbManager>,
+
+    /// Tracker used to remember the current consensus state.
+    state_tracker: state_tracker::StateTracker<D>,
+
+    /// Broadcast channel used to publish state updates.
+    cupdate_tx: broadcast::Sender<Arc<ClientUpdateNotif>>,
+}
+
+impl<D: Database> WorkerState<D> {
+    /// Constructs a new instance by reconstructing the current consensus state
+    /// from the provided database layer.
+    pub fn open(
+        params: Arc<Params>,
+        database: Arc<D>,
+        l2_block_manager: Arc<L2BlockManager>,
+        cupdate_tx: broadcast::Sender<Arc<ClientUpdateNotif>>,
+        checkpoint_manager: Arc<CheckpointDbManager>,
+    ) -> anyhow::Result<Self> {
+        let cs_prov = database.client_state_provider().as_ref();
+        let (cur_state_idx, cur_state) = state_tracker::reconstruct_cur_state(cs_prov)?;
+        let state_tracker = state_tracker::StateTracker::new(
+            params.clone(),
+            database.clone(),
+            cur_state_idx,
+            Arc::new(cur_state),
+        );
+
+        // TODO make configurable
+        let config = CsmExecConfig {
+            retry_base_dur: time::Duration::from_millis(1000),
+            // These settings makes the last retry delay be 6 seconds.
+            retry_cnt_max: 20,
+            retry_backoff_mult: 1120,
+        };
+
+        Ok(Self {
+            params,
+            config,
+            database,
+            l2_block_manager,
+            state_tracker,
+            cupdate_tx,
+            checkpoint_manager,
+        })
+    }
+
+    /// Gets the index of the current state.
+    pub fn cur_event_idx(&self) -> u64 {
+        self.state_tracker.cur_state_idx()
+    }
+
+    /// Gets a ref to the consensus state from the inner state tracker.
+    pub fn cur_state(&self) -> &Arc<ClientState> {
+        self.state_tracker.cur_state()
+    }
+
+    /// Gets a reference to checkpoint manager
+    pub fn checkpoint_db(&self) -> &CheckpointDbManager {
+        self.checkpoint_manager.as_ref()
+    }
+}
+
+/// Receives messages from channel to update consensus state with.
+// TODO consolidate all these channels into container/"io" types
+pub fn client_worker_task<D: Database, E: ExecEngineCtl>(
+    shutdown: ShutdownGuard,
+    mut state: WorkerState<D>,
+    engine: Arc<E>,
+    mut msg_rx: mpsc::Receiver<CsmMessage>,
+    status_tx: Arc<StatusTx>,
```

Comment:
same as above

Reply:
This was just moving code around that already exists.